adcc/adc_pp/matrix.py

```diff
@@ -0,0 +1,565 @@
+        #      which is the right thing to do, but not yet supported.
+        # + 0.5 * einsum("icjd,klac,klbd->iajb", hf.ovov, mp.t2oo, mp.t2oo)
+        + 0.5 * einsum("icjd,acbd->iajb", hf.ovov,
+                       einsum("klac,klbd->acbd", mp.t2oo, mp.t2oo))
+        + 0.5 * einsum("ikcd,jlcd,kalb->iajb", mp.t2oo, mp.t2oo, hf.ovov)
+        - einsum("iljk,kalb->iajb", hf.oooo, t2sq)
+        - einsum("idjc,acbd->iajb", t2sq, hf.vvvv)
+    )
+
+
+@register_as_intermediate
+def cvs_adc3_m11(hf, mp, intermediates):
+    i1 = adc3_i1(hf, mp, intermediates).evaluate()
+    i2 = cvs_adc3_i2(hf, mp, intermediates).evaluate()
+    t2sq = einsum("ikac,jkbc->iajb", mp.t2oo, mp.t2oo).evaluate()
+
+    # Build two Kronecker deltas
+    d_cc = zeros_like(hf.fcc)
+    d_vv = zeros_like(hf.fvv)
+    d_cc.set_mask("II", 1.0)
+    d_vv.set_mask("aa", 1.0)
+
+    return (
+        + einsum("IJ,ab->IaJb", d_cc, hf.fvv + i1)
+        - einsum("IJ,ab->IaJb", hf.fcc - i2, d_vv)
+        - einsum("JaIb->IaJb", hf.cvcv)
+        + (  # symmetrise I<>J and a<>b
+            + einsum("JaIc,bc->IaJb", hf.cvcv, intermediates.cvs_p0_vv)
+            - einsum("kIJa,kb->IaJb", hf.occv, 2.0 * intermediates.cvs_p0_ov)
+        ).symmetrise((0, 2), (1, 3))
+        # TODO This hack is done to avoid opt_einsum being smart and instantiating
+        #      a tensor of dimension 6 (to avoid the vvvv tensor) in some cases,
+        #      which is the right thing to do, but not yet supported.
+        # + 0.5 * einsum("IcJd,klac,klbd->IaJb", hf.cvcv, mp.t2oo, mp.t2oo)
+        + 0.5 * einsum("IcJd,acbd->IaJb", hf.cvcv,
+                       einsum("klac,klbd->acbd", mp.t2oo, mp.t2oo))
+        - einsum("lIkJ,kalb->IaJb", hf.ococ, t2sq)
+    )
+
+
+@register_as_intermediate
+def adc3_pia(hf, mp, intermediates):
+    # Note that this definition differs from libadc, since it includes
+    # the hf.ooov term as well.
+    return (                          # Perturbation theory in ADC coupling block
+        + hf.ooov                                            # 1st order
+        - 2.0 * mp.t2eri(b.ooov, b.ov).antisymmetrise(0, 1)  # 2nd order
+        - 0.5 * mp.t2eri(b.ooov, b.vv)                       # 2nd order
+    )
+
+
+@register_as_intermediate
+def cvs_adc3_pia(hf, mp, intermediates):
+    # Perturbation theory in CVS-ADC coupling block:
+    #       1st                     2nd
+    return hf.occv - einsum("jlac,lKIc->jIKa", mp.t2oo, hf.occv)
+
+
+@register_as_intermediate
+def adc3_pib(hf, mp, intermediates):
+    # Note that this definition differs from libadc, since it includes
+    # the hf.ovvv term as well.
+    return (                          # Perturbation theory in ADC coupling block
+        + hf.ovvv                                            # 1st order
+        + 2.0 * mp.t2eri(b.ovvv, b.ov).antisymmetrise(2, 3)  # 2nd order
+        - 0.5 * mp.t2eri(b.ovvv, b.oo)                       # 2nd order
+    )
```

Comment:
I'm impressed... this is almost perfection 👍 
Very clean and the block dispatch is really cool! 🚀